### PR TITLE
add support for multi-threaded `west update`

### PR DESCRIPTION
**Pull request by M1cha**

### CODE REVIEW

**Comment by M1cha** [comment]
does anyone know why tox can't find `types-PyYaml`?
I have the same error locally even though it is installed.

**Comment by mbolivar-nordic** [comment]
Thanks for the PR!

> does anyone know why tox can't find types-PyYaml?

tox runs tests in its own virtual environment. This is important for making sure we've described the package dependencies correctly. Check this commit in the zephyr repository for information on how to fix: https://github.com/zephyrproject-rtos/zephyr/commit/8875340db4dd55ad89e174e7124b7cfb2b3db717

If you can prepend a commit onto this PR fixing that, it'd be much appreciated, thanks.

**Review by mbolivar-nordic — COMMENTED** [review]
Thanks! I've been hoping to find time to add this feature for a while, and I'm really delighted you beat me to it. I would like to merge it, but I have one request.

This new `west update -j N` behaves like `repo sync -j N` in that the git output for the different updates are all commingled into stdout and stderr. This leads to confusing things like this in my testing:

```

=== updating hal_altera (modules/hal/altera):
=== updating cmsis (modules/hal/cmsis):
=== updating civetweb (modules/lib/civetweb):
=== updating fatfs (modules/fs/fatfs):
=== updating hal_infineon (modules/hal/infineon):
```

Followed by a sequence of `HEAD is now at` messages for a single project, like this:

```
=== updating hal_espressif (modules/hal/espressif):
HEAD is now at b4d31f3 hal: nuvoton: add license to numicro m480 series hal
HEAD is now at 81a059f Merge pull request #1 from nandojve/pdl_scb_spi_i2c
```

I know people are opting in to this by using `-j`, but I think this is unnecessarily confusing.

Could you instead please store the output from each git subprocess and print the output results from each project in a self-contained way from the main thread?

**Comment by M1cha** [comment]
```
Could you instead please store the output from each git subprocess and print the output results from each project in a self-contained way from the main thread?
```
That would only work for a full summary at the end because if you'd delay them until the end you loose the live-status updates. this would be very annoying for huge projects which can take hours to download.

**Comment by mbolivar-nordic** [comment]
That's a fair point, but I'm hoping it's solvable. [jiri update](https://fuchsia.googlesource.com/jiri/) does not seem to have this problem, and it appears to use cursor positioning commands in the terminal to print update progress progressively.

West already has a colorama dependency, which provides cross-platform cursor positioning; this should work even on Windows according to its docs:

https://github.com/tartley/colorama/blob/master/demos/demo06.py

I think it's worth solving this problem before introducing this feature; many west features were developed out of frustration with repo's limitations, and I'd like to continue to do things in a nice way.

Do you think you give this a go?

**Comment by M1cha** [comment]
I just tried jiri and they do seem to capture all output but they don't show any live git-output, so apparently they didn't have to solve that by not having that feature.
Technical possibilities aside, I'm not sure what I'd expect the output to look like.
We can't really do our own detailed progressbar without using libgit2 instead of the git cli (or worse: parsing git CLI output).
so the only thing we could do would be something like terminal emulators like tmux do: record the full log of all jobs and fully reset the viewport to show the output of a specific process. Then, we could periodically cycle through them (instead of offering a manual switch) so you have a chance to see the status of each.

Another important question is what all of that will look like on a CI like github actions without an interactive TTY.
Do we just accept scrambled output in CI logs or do we want to hold logs back like jiri and maybe print the full log after each process is done?

**Comment by mbolivar-nordic** [comment]
>  We can't really do our own detailed progressbar without using libgit2 instead of the git cli (or worse: parsing git CLI output).

That's fair. Let's not get into parsing git CLI output (that would be horrible), and libgit2 is likely overkill.

> so the only thing we could do would be something like terminal emulators like tmux do

I'm just brainstorming here since I don't have as much experience in more sophisticated terminal UIs as I'd like (which is part of why I didn't tackle this change).

Would it be possible to have the output for the commands which have been completed get printed and allowed to scroll off the terminal as usual, but also print the final line or two of output from ongoing updates printed to the bottom of the terminal?

Something like this:

```

=== updating hal_ti (modules/hal/ti):
HEAD is now at 3da6fae zephyr: use k_ms_to_ticks_ceil32 instead of z_ms_to_ticks
=== updating hal_xtensa (modules/hal/xtensa):
HEAD is now at 2f04b61 zephyr: add SoC overlay for intel_tgl_adsp
=== updating libmetal (modules/hal/libmetal):
HEAD is now at 39d049d lib: update libmetal to release v2021.04.0 + VERSION path fix
=== updating littlefs (modules/fs/littlefs):
HEAD is now at 9e4498d lfs_util: fix log declaration when logging is disabled
=== updating loramac-node (modules/lib/loramac-node):
HEAD is now at 12019623 zephyr: radio: sx127x: override buffer read/write functions
=== updating lvgl (modules/lib/gui/lvgl):
HEAD is now at 31acbaa3 lv_font: Fix compiler warning when compressed fonts are disabled (#1880)
=== updating lz4 (modules/lib/lz4):
HEAD is now at 8e303c2 zephyr: make this repo a zephyr module
=== updating mcuboot (bootloader/mcuboot):
HEAD is now at 7a51968 Merge: synhronize up to https://github.com/mcu-tools/mcuboot/commit/680ed07e42f

^^^ finished lines are scrolling off the top

=== updating tensorflow (modules/lib/tensorflow):
Receiving objects:   1% (12833/1137494), 3.91 MiB | 678.00 KiB/s 

^^^ just a line (or a few lines) of output for projects which are still updating

```

Please forgive me if that's a naive question.


>  Do we just accept scrambled output in CI logs or do we want to hold logs back like jiri and maybe print the full log after each process is done?

I think the latter, yeah; basically save the entire output `if not sys.stdout.isatty()` or so.

**Comment by marc-hb** [comment]
Is there some specific reason why the `class Update` was moved to a different place in the file? This creates a huge diff much harder to review.


> That would only work for a full summary at the end because if you'd delay them until the end you loose the live-status updates. That would be very annoying for huge projects which can take hours to download.

Not sure I understand this... you don't have to delay all outputs till the end, you could "simply" print the complete output of each git process as soon as it's finished, something like:

```
Updating 30/30 repos
=== updating hal_ti (modules/hal/ti):
HEAD is now at 3da6fae zephyr: use k_ms_to_ticks_ceil32 instead of z_ms_to_ticks
Updating 29/30 repos
=== updating hal_xtensa (modules/hal/xtensa):
HEAD is now at 2f04b61 zephyr: add SoC overlay for intel_tgl_adsp
Updating 28/30 repos
=== updating libmetal (modules/hal/libmetal):
HEAD is now at 39d049d lib: update libmetal to release v2021.04.0 + VERSION path fix
```

I see you're collecting jobs with `multiprocessing.pool.Pool.starmap` right now, does that block until all jobs are complete?

Here's a piece of Python that submits all jobs at once and collects them one by one as soon as they're completed:
https://gist.github.com/marc-hb/b537d30fd3d4fc0d506eff1b66f745e6

**Comment by marc-hb** [comment]
Just thought about this: one area I think needs some test attention is aborting the update with Ctrl-C or similar. Not sure how that was  implemented in the past but you want to make sure the thread cancellation is "not too brutal" and does not leave corrupted git repos behind.

**Comment by M1cha** [comment]
> Would it be possible to have the output for the commands which have been completed get printed and allowed to scroll off the terminal as usual, but also print the final line or two of output from ongoing updates printed to the bottom of the terminal?

Without being a terminal pro myself, yes I think this is possible.

> Is there some specific reason why the class Update was moved to a different place in the file? This creates a huge diff much harder to review.

I touched on that in the commit message of 63fec8f9b1366ab3ed0268d2b061ce4baa9345c2
the Pool creates new separate processes so it has to serialize and deserialize worker arguments.
the argparse object is not serializable though. Since we never change the args I wonder if we could store `self` in a global variable so it gets copied during `fork` :thinking: 

 > Not sure I understand this... you don't have to delay all outputs till the end, you could "simply" print the complete output of each git process as soon as it's finished, something like:

Even a single project can take an hour to download(e.g. the linux kernel). You wouldn't have any feedback whatesoever during that time.

> I see you're collecting jobs with multiprocessing.pool.Pool.starmap right now, does that block until all jobs are complete?

yes

**Comment by marc-hb** [comment]
> Even a single project can take an hour to download(e.g. the linux kernel). You wouldn't have any feedback whatesoever during that time.

In that case does west give any now?

**Comment by marc-hb** [comment]
I think the feature itself is too valuable to be delayed by these "progress bar" issues that no other similar tool seems to have solved yet - maybe for good reasons like... it's hard? So how about this: for now add support for two limited by doable display modes:
- jumbled up mode like git-repo = what this does already (I didn't fetch and try it, merely relying on @mbolivar-nordic 's description
- Report-each-completion mode that I suggested earlier. Does not solve the problem of huge repos like the Linux  kernel but provide "some" feedback as long as some repos are small. Note this is a problem only when cloning for the first time.

Micro-optimization: deliberately stagger the fetch commands slightly to give an edge to some repos and increase the chances that feedback is more spread in time. Also avoids triggering checks for DoS attacks and other abuse and  gives ssh ControlMaster connections more time to complete.

Room for future improvement: be the first multi-repo tool to implement a parallel progress  bar.

One million $ question: which mode should be the default? Probably the jumbled up mode in interactive use (with a warning and invite to use the other one) to make sure interactive users don't think something is stuck. Report-each-completion mode by default when not TTY (e.g. CI).

The code for the two modes should barely differ from each other (but it would differ from the current one)

**Comment by marc-hb** [comment]
> which mode should be the default? Probably the jumbled up mode in interactive use ...

I meant jumbled up by default _if_ the user uses `-jN` with N>2. The default-default should be `-j1` as it is now, so no ugly output by default-default.

**Comment by M1cha** [comment]
I'm not sure what the best way would be to capture stderr and stdout in a platform-independent way.
`multiprocessing.Pool` does create actual processes but doesn't seem to provide any capturing functionality like `subprocess`.
In addition to that `subprocess` seams to bypass `sys.stdout` so even if I override that, the output of those is still being shown.

Now, I could just do it the C-way and use `dup` to replace stdout and stderr with a file to capture both python and subprocess logs at the same time but I doubt that'd work on windows.

**Comment by pearson** [comment]
I've got to admit that I'm impressed by this PR. I just did a new install of Zephyr and a `west update -j 20` took at most a minute to complete. 

As mentioned above by @marc-hb , this feature is far too useful to delay due to the "intermingling" of the console output. That might be an issue if some level of job threading became the default (which I'd argue it should), but as an optional feature it does not seem to be problem.

@M1cha , if you want to test the "C-way" on a Windows box, feel free to ping me in the Slack channel. I'd be happy to help out!

**Comment by sjg20** [comment]
From my POV I don't want any output from such a routine task. I wish the tools were less verbose and that includes the build system. Just tell me about warnings and errors, unless I enable some sort of debug mode.

So:

$ west update
$

You can put the output in a log file perhaps, and show any errors, but I just don't see the point of spitting out lots of junk.

**Comment by pearson** [comment]
@sjg20, it seems to be exactly the same output one would normally see from West. The only difference is that, should multiple job threads be used, their output can end up intermingled. 

For example, if two git jobs are running at the same time, their status bar updates overlap in the console output.

**Comment by marc-hb** [comment]
> I wish the tools were less verbose and that includes the build system.

You can already do that now: just redirect stderr to a log file or `/dev/null` / `Out-Null` when running west. Git will detect that this is not a terminal and will not emit any progress information (not even to your log file). Git will still print basic information on stdout. To be clear: for now `west` just shares stdout and stderr "as is" with git and west has NO involvement in what I just described.

The problem with suppressing progress bars likes this is that git output now looks exactly the same as when you have a proxy or firewall misconfiguration or network outage. Maybe never a problem for you but many other people less lucky with their network do care about progress bars; that's why they exist. This being said and repeating myself: the lack of the perfect progress bars should not delay this awesome feature. Just don't turn the feature by default and show some warnings about bad or no progress bars when it's explicitly requested. "Release early, release often".

**Comment by mbolivar-nordic** [comment]
Thanks for patience while I was on vacation.

I am convinced by @marc-hb that this is not worth delaying and that we should have two modes of operation ('unbuffered' as it is now, and 'print the output of each command after it completes' as I originally suggested.

I would like the 'print the output of each command after it completes' be the default though. Can we have the other mode of behavior enabled by an `--unbuffered` flag option? Thoughts on that?

**Comment by mbolivar-nordic** [comment]
> From my POV I don't want any output from such a routine task. I wish the tools were less verbose and that includes the build system. Just tell me about warnings and errors, unless I enable some sort of debug mode.

This feels like an unrelated request -- #301

**Comment by sjg20** [comment]
@marc-hb no that is not the same thing. Tools should not spit out junk when they are working properly, only warnings and errors, things which are actionable for the user. Provide a verbose/debug mode if needed.

@mbolivar-nordic OK I added a thought to that issue. My point for this one is that it doesn't seem that important the ordering of the output.

**Comment by M1cha** [comment]
I pushed my current code but it's still heavily WIP due to log capturing which seems to be really hard to do in python.
Using os.dup2 seems to work just fine on all operating systems. The problem passing data from the host process to the child processes for operating systems which use spawn instead of fork. Neither the update class nor the multiprocess lock are pickleable right now.

I have to think about this a lot more so it's gonna be way more work than expected. I'll keep you updated.

**Comment by M1cha** [comment]
by hackishly settings self.parser to None i made the update object picklable.
I removed the need for a lock by using named files inside a temporary directory.

This solution seems to work on all operating systems as far as the tests are concerned.

**Comment by mbolivar-nordic** [comment]
Thanks @M1cha. I tested it out and took a look and 'scrambled' output is still the default, so this request is still open:

> I would like the 'print the output of each command after it completes' be the default though.

Can you please update (heh) this?

**Review by mbolivar-nordic — APPROVED** [review]
This is working well in my limited testing, thanks. I'm going to be cutting another west release very soon to fix a bug reported elsewhere, so I'm just going to merge this and give it some more testing on master before the release. We can address any bugs that turns up later. Thanks again for contributing this; it's much appreciated!

**Comment by mbolivar-nordic** [comment]
BTW, if you want to send a follow up commit adding an `update.unbuffered` config option that can be set to `true` to make `--unbuffered` the default, please feel free!

**Comment by mbolivar-nordic** [comment]
>  I'm just going to merge this

... if CI ever finishes, that is...

**Comment by M1cha** [comment]
> BTW, if you want to send a follow up commit adding an `update.unbuffered` config option that can be set to `true` to make `--unbuffered` the default, please feel free!

wasn't your whole point that you don't want it to be the default? :thinking:

**Comment by mbolivar-nordic** [comment]
> wasn't your whole point that you don't want it to be the default? thinking

What I mean is, if you want users to be able to do this:

```
$ west config update.unbuffered true
$ west update -j8 # this is unbuffered
$ west config -d update.unbuffered
$ west update -j8 # this is buffered
```

in order to be able to configure your preferred behavior, I'd be happy to merge that too.

**Comment by M1cha** [comment]
ah yea that's a good idea, will do :)

File: src/west/app/project.py
Comment (marc-hb) [inline src/west/app/project.py:912]: It seems weird and very unusual to redirect stdout to stderr and it feels like this could be a cause for false alarms. Please add a comment explaining why you do that, probably because git (strangely?) prints a lot of harmless  info to stderr and ( if you know why git does that then even better: say why!).

**Review by marc-hb — DISMISSED** [review]
Extremely clear and concise code, well done! A bit _too_ concise actually :-) Outside the test code which is heavily commented, I found zero comments and no `log.dbg()` either! I made a number of suggestions below which would have saved me some cognitive load and would have sped up my review significantly.

To be clear: there is no single reason why I'm "Requesting Changes" and please disagree with some suggestions.  I'm requesting changes only because of the number of suggestions I made and the complete lack of comments, pydocs and log statements.

**Review by marc-hb — COMMENTED** [review]
> BTW, if you want to send a follow up commit ...

That works too.

**Comment by mbolivar-nordic** [comment]
> Why are there 3 different places to dismiss a review, out of sync with each other?

Your earlier comment makes it seem like you wanted to request changes before I merge. That's fine, but if so, can you please click the 'request changes' button since I've approved this already? If not, just let me know and I'll merge it when the last macos test run in CI goes green.

**Review by marc-hb — CHANGES_REQUESTED** [review]
> Your earlier comment makes it seem like you wanted to request changes before I merge. That's fine, but if so, can you please click the 'request changes' button since I've approved this already?

I did request changes (mostly: comments) but then I thought maybe you want to make a release before @M1cha even gets a chance to make any change. As you confirmed there is some time left, re-requesting changes again.

File: src/west/app/project.py
Comment (M1cha) [reply src/west/app/project.py:896]: will do, keep in mind that:
- the directory get deleted as soon as `logdir` goes out og scope
- the logs may not appear in the file depending on both python and kernel buffering.

Comment (M1cha) [reply src/west/app/project.py:896]: I'll add the isatty info to the help of `--jobs` because that seems more important to the user of that option than it is to the reader of the code.

File: src/west/app/project.py
Comment (BeckmaR) [inline src/west/app/project.py:869]: I would remove the outer try, I don't think it has any purpose here. A single try / except / finally should do the same thing.

File: src/west/app/project.py
Comment (marc-hb) [reply src/west/app/project.py]: > the directory get deleted as soon as logdir goes out og scope
> the logs may not appear in the file depending on both python and kernel buffering.

Understood. I think the most common reason why someone will want this information is because the download silently hangs because firewall/proxy/etc.

**Review by marc-hb — APPROVED** [review]
Well done, thank you!

**Review by mbolivar-nordic — APPROVED** [review]

**Comment by mcuxcc** [comment]
I am not sure if it is correct to comment on this thread, but can anyone help me to understand why this patch not in main branch?

**Comment by M1cha** [comment]
@mcuxcc it got reverted because of broken windows support. I intended to fix it but lost the motivation to do so myself.

**Comment by mcuxcc** [comment]
Recently we are integrating west, but we found the performance issue in CI
so  i am looking for such solutions

**Comment by marc-hb** [comment]
See
- #551

**Pull request merged**
